### PR TITLE
Refactor mempool spend conflict checks to increase performance

### LINKED ISSUE #2784: Refactor spend conflict checks in the mempool storage to increase performance
## Motivation

ZIP-401 prevents memory DoS by setting a maximum mempool size. This mempool size is much larger than Zebra's current mempool size.

PR #2765 implemented a simple algorithm to check if a transaction has spend conflicts with another transaction already in the mempool.

That algorithm works well for a mempool with few transactions. But as the size of the mempool increases, performance will be impacted. The algorithm is `O(n*m)`, where `n` is the number of spent UTXOs and revealed nullifiers in the transaction and `m` is the number of spent UTXOs and revealed nullifiers in all transactions already in the mempool.

So as part of implementing ZIP-401, we need to make the spend conflict checks more efficient. Otherwise, we replace a memory DoS with a CPU DoS.

### Specifications

> Adoption of this proposal would increase robustness of Zcash nodes against denial-of-service attack, in particular attacks that attempt to exhaust node memory.

https://zips.z.cash/zip-0401#motivation

> The default value of 80000000 for `mempooltxcostlimit` represents no more than 40 blocks’ worth of transactions in the worst case, which is the default expiration height after the Blossom network upgrade.

https://zips.z.cash/zip-0401#rationale

### Designs

The original idea was to to cache the spent UTXOs and revealed nullifiers from all transactions in
the mempool using `HashSet`s. This would mean that the conflict check would only be `O(n)`, but it adds complexity to the code to add a transaction to the mempool and to remove a transaction from the mempool, because the caches must be updated.

If this idea is implemented, the `has_spend_conflicts` method could be updated to something like:

```rust
fn has_spend_conflicts<'slf, 'tx, Extractor, Outputs, Output>(
    &'slf self,
    tx: &'tx UnminedTx,
    extractor: Extractor,
    existing_outputs: &'slf HashSet<Output>,
) -> bool
where
    'slf: 'tx,
    Extractor: Fn(&'tx Transaction) -> Outputs,
    Outputs: IntoIterator<Item = Output>,
    Output: Eq + Hash + 'tx,
{
    let new_outputs = extractor(&tx.transaction).into_iter();

    new_outputs.any(|output| existing_outputs.contains(output))
}
```

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

#2765 implemented rejection of mempool transactions that have spend conflicts (i.e., spend the same transparent outpoint or reveal the same nullifier) with other transactions already in the mempool. However, the implementation was simple and would not scale well in terms of performance as the mempool size is increased.

## Solution

Replace the O(n * m) algorithm with an O(n) algorithm. Now the `mempool::Storage` uses `HashSet`s to keep track of the spent outpoints and revealed nullifiers by the verified transactions. This allows it to quickly check if there are conflicts when new transactions are added.

## Review

@teor2345 

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebrad/src/components/mempool/storage.rs [hunk 1]
```diff
@@ -42,8 +44,21 @@ pub struct Storage {
...
+    spent_outpoints: HashSet<transparent::OutPoint>,
+
+    /// The set of revealed Sprout nullifiers.
+    sprout_nullifiers: HashSet<sprout::Nullifier>,
+
+    /// The set of revealed Sapling nullifiers.
+    sapling_nullifiers: HashSet<sapling::Nullifier>,
+
+    /// The set of revealed Orchard nullifiers.
+    orchard_nullifiers: HashSet<orchard::Nullifier>,
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs]:
```suggestion
    /// The set of Sprout nullifiers revealed by mempool transactions.
    sprout_nullifiers: HashSet<sprout::Nullifier>,

    /// The set of Sapling nullifiers revealed by mempool transactions.
    sapling_nullifiers: HashSet<sapling::Nullifier>,

    /// The set of Orchard nullifiers revealed by mempool transactions.
    orchard_nullifiers: HashSet<orchard::Nullifier>,
```

File: zebrad/src/components/mempool/storage.rs [hunk 2]
```diff
@@ -156,39 +175,82 @@ impl Storage {
...
 
-        self.verified
-            .iter()
-            .flat_map(|tx| extractor(&tx.transaction))
-            .any(|output| new_outputs.contains(&output))
+        for new_output in new_outputs {
+            if existing_outputs.insert(new_output.clone().into_owned()) {
+                inserted_outputs.push(new_output);
+            } else {
+                Self::remove_from_set(existing_outputs, inserted_outputs);
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs]: Likely Bug: If the conflict happens in the sprout nullifiers, are the UTXOs from the conflicting transaction still added to `spent_outpoints`?

I suggest a two-phase approach, where you first check for any conflicts across all sets, then update all sets.

Can you also add a test case that covers this situation?

File: zebrad/src/components/mempool/storage.rs [hunk 3]
```diff
@@ -42,8 +44,21 @@ pub struct Storage {
     /// The set of verified transactions in the mempool. This is a
     /// cache of size [`MEMPOOL_SIZE`].
     verified: VecDeque<UnminedTx>,
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs]: Design: It might be helpful to turn `verified` and its associated spent sets into an inner type, so that you can make sure that inserts and removals update them all.

File: zebrad/src/components/mempool/storage.rs [hunk 4]
```diff
@@ -88,8 +103,11 @@ impl Storage {
         // Once inserted, we evict transactions over the pool size limit in FIFO
         // order.
         if self.verified.len() > MEMPOOL_SIZE {
-            for evicted_tx in self.verified.drain(MEMPOOL_SIZE..) {
+            let newly_evicted: Vec<_> = self.verified.drain(MEMPOOL_SIZE..).collect();
+
+            for evicted_tx in newly_evicted {
                 let _ = self.rejected.insert(evicted_tx.id, State::Excess);
+                self.remove_outputs(&evicted_tx);
             }
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs]: Bug Risk: Can we just use `remove` here?

That way, we aren't trying to keep `remove` and this code in sync.

Some of my changes in PR #2827 might help here:
https://github.com/ZcashFoundation/zebra/pull/2827/files#diff-718f7dc7ea29bde23b4ccf94195f207750fe7d333dcb1c820195267bc9feada2R122

**Review by teor2345 — COMMENTED** [review]
This looks good, but I think we need to be really careful to avoid modifying anything when there's a conflict.

(I'm sorry I didn't submit this yesterday, I was a bit distracted by the release.)

File: zebrad/src/components/mempool/storage.rs (hunk 3 above)
Comment (jvff) [reply zebrad/src/components/mempool/storage.rs]: I tried implementing this, but I'm not sure if I'm on the right path :thinking:

File: zebrad/src/components/mempool/storage.rs (hunk 4 above)
Comment (jvff) [reply zebrad/src/components/mempool/storage.rs]: I think I managed to do this, but I'm using the `VecDeque` index for removal. I'm wondering if we'll have to change from a `usize` index into a `UnminedTxId` later on :/

File: zebrad/src/components/mempool/storage.rs (hunk 2 above)
Comment (jvff) [reply zebrad/src/components/mempool/storage.rs]: Done, but I'll have to add a test tomorrow :sweat_smile:

File: zebrad/src/components/mempool/storage.rs (hunk 4 above)
Comment (teor2345) [reply zebrad/src/components/mempool/storage.rs]: >  I'm wondering if we'll have to change from a `usize` index into a `UnminedTxId` later on :/

That's possible in future, but ZIP 401 is the only big chunk of work we have left. And I can't see anything in ZIP 401 that requires eviction by ID. (It only does eviction by weight.)

File: zebrad/src/components/mempool/storage.rs [hunk 5]
```diff
@@ -69,43 +71,44 @@ impl Storage {
...
-        // TODO: use random weighted eviction as specified in ZIP-401 (#2780)
-        if self.verified.len() > MEMPOOL_SIZE {
-            for evicted_tx in self.verified.drain(MEMPOOL_SIZE..) {
-                let _ = self
-                    .rejected
-                    .insert(evicted_tx.id, StorageRejectionError::RandomlyEvicted);
-            }
-
-            assert_eq!(self.verified.len(), MEMPOOL_SIZE);
+        // This will a evict transactions to open space for the new transaction if needed.
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs]: Confusing comment: I'm wondering if this comment is outdated, or the code wasn't updated.

The next code block explicitly does the eviction, rather than doing it inside `verified.insert()`.

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+            Some(self.remove(last_index))
+        }
+    }
+
+    /// Removes all transactions in the set that matches the `predicate`.
+    ///
+    /// Returns the amount of transactions removed.
+    pub fn remove_all_that(&mut self, predicate: impl Fn(&UnminedTx) -> bool) -> usize {
+        // Clippy is unable to detect that there will be a borrow conflict without the `collect`.
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs]: Nitpick: Typo?
```suggestion
        // Clippy is unable to detect that there won't be a borrow conflict without the `collect`.
```

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+    /// Insert a transaction into the set.
+    ///
+    /// # Correctness
+    ///
+    /// The transaction is assumed to not have spend conflicts with any other transaction already
+    /// in the set. Use [`Self::has_spend_conflicts`] to ensure this before calling this method.
+    pub fn insert(&mut self, transaction: UnminedTx) {
+        self.cache_outputs_from(&transaction.transaction);
+        self.transactions.push_front(dbg!(transaction));
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs]: Nitpick: Leftover `dbg!`
```suggestion
        self.transactions.push_front(transaction);
```

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+            self.remove(index_to_remove);
+        }
+
+        removed_count
+    }
+
+    /// Removes a transaction from the set.
+    ///
+    /// Also removes its outputs from the internal caches.
+    // TODO: Maybe replace `usize` with `UnminedTxId` when `VecDeque` is no longer used?
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs]: Potential Scope Creep: Can we either link to a ticket where this change is going to happen, or remove this comment to avoid scope creep?
```suggestion
```

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+            .transactions
+            .iter()
+            .enumerate()
+            .filter(|(_, tx)| predicate(tx))
+            .map(|(index, _)| index)
+            .collect();
+
+        let removed_count = indices_to_remove.len();
+
+        for index_to_remove in indices_to_remove.into_iter().rev() {
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs:121]: Subtle, Correctness-Critical Code: Ordering is important here, but it's not immediately clear why.

Please also add a test that removes multiple transactions at the same time, and makes sure the correct transactions were removed. You might need to increase the mempool size to 3 to do this 🤣 

```suggestion
        // Correctness: remove indexes in reverse order,
        // so earlier indexes still correspond to the same transactions
        for index_to_remove in indices_to_remove.into_iter().rev() {
```

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        self.sapling_nullifiers.clear();
+        self.orchard_nullifiers.clear();
+    }
+
+    /// Insert a transaction into the set.
+    ///
+    /// # Correctness
+    ///
+    /// The transaction is assumed to not have spend conflicts with any other transaction already
+    /// in the set. Use [`Self::has_spend_conflicts`] to ensure this before calling this method.
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs]: Preventing API Misuse: Can we do the `has_spend_conflicts` check inside `insert`, and return a `Result`?

We can re-use the existing spend conflict error type.

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+    pub fn insert(&mut self, transaction: UnminedTx) {
+        self.cache_outputs_from(&transaction.transaction);
+        self.transactions.push_front(dbg!(transaction));
+    }
+
+    /// Evict one transaction from the set to open space for another transaction.
+    pub fn evict_one(&mut self) -> Option<UnminedTx> {
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs:89]: This is a nice API!

File: zebrad/src/components/mempool/storage/verified_set.rs [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+            None
+        } else {
+            // TODO: use random weighted eviction as specified in ZIP-401 (#2780)
+            let last_index = self.transactions.len() - 1;
+
+            Some(self.remove(last_index))
+        }
+    }
+
+    /// Removes all transactions in the set that matches the `predicate`.
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/verified_set.rs]: Nitpick: Typo
```suggestion
    /// Removes all transactions in the set that match the `predicate`.
```

**Review by teor2345 — COMMENTED** [review]
I really like this refactor, and we could merge it as-is.
(After we add some missing tests.)

I have a few optional suggestions for improvements.

File: zebrad/src/components/mempool/storage/verified_set.rs (hunk 10 above)
Comment (teor2345) [reply zebrad/src/components/mempool/storage/verified_set.rs]: I also think we'll need to return a `Result` so we know whether to gossip the transaction:
https://github.com/ZcashFoundation/zebra/pull/2823#discussion_r721815903

Comment (jvff) [reply zebrad/src/components/mempool/storage/verified_set.rs]: Done.

File: zebrad/src/components/mempool/storage/verified_set.rs (hunk 6 above)
Comment (jvff) [reply zebrad/src/components/mempool/storage/verified_set.rs]: I think there's no typo, but the wording was a bit confusing, so I tried to make it a bit clearer.

File: zebrad/src/components/mempool/storage.rs (hunk 5 above)
Comment (jvff) [reply zebrad/src/components/mempool/storage.rs]: Done. Yeah, that was leftover code from a different attempt I did :sweat_smile:

File: zebrad/src/components/mempool/storage.rs (hunk 2 above)
Comment (jvff) [reply zebrad/src/components/mempool/storage.rs]: Okay, this test ended up being more complicated than I expected. It feels far from my best work, but hopefully it's good enough :sweat:

**Comment by conradoplg** [comment]
I reviewed this and it looks good, but I think it's better for @teor2345 to take a last look. I can make any final adjustments if needed.

File: zebrad/src/components/mempool/storage/verified_set.rs (hunk 9 above)
Comment (teor2345) [reply zebrad/src/components/mempool/storage/verified_set.rs:121]: The test works, but only if the mempool size is at least 4.
(It could theoretically work with 3, but failures are rare at that size.)

I increased the mempool size, and committed a proptest seed that fails if we remove transactions in ascending index order.

**Review by teor2345 — APPROVED** [review]
I increased the mempool size to 4, so that the new tests would actually detect multiple removal errors.

I also summarised the debug output of the tests, because it was very very large.
(It contained the entire contents of a vector of random transactions.)

Apart from the mempool size, I only changed test debug formatting.
So I think we should merge now to avoid conflicts.

**Review by dconnolly — COMMENTED** [review]
🤩
